linuxkm/lkcapi_glue.c

```diff
@@ -0,0 +1,2537 @@
+            pr_err("skcipher_walk_done() failed: %d\n", err);
+            return err;
+        }
+    }
+
+    return err;
+}
+
+static int km_AesXtsDecrypt(struct skcipher_request *req)
+{
+    int                      err = 0;
+    struct crypto_skcipher * tfm = NULL;
+    struct km_AesXtsCtx *    ctx = NULL;
+    struct skcipher_walk     walk;
+    unsigned int             nbytes = 0;
+
+    tfm = crypto_skcipher_reqtfm(req);
+    ctx = crypto_skcipher_ctx(tfm);
+
+    err = skcipher_walk_virt(&walk, req, false);
+
+    if (unlikely(err)) {
+        pr_err("skcipher_walk_virt() failed: %d\n", err);
+        return err;
+    }
+
+    while ((nbytes = walk.nbytes)) {
+        err = wc_AesXtsDecrypt(ctx->aesXts, walk.dst.virt.addr,
+                               walk.src.virt.addr, nbytes,
+                               walk.iv, walk.ivsize);
+
+        if (unlikely(err)) {
+            pr_err("wc_AesCbcDecrypt failed");
+            return err;
+        }
+
+        err = skcipher_walk_done(&walk, walk.nbytes - nbytes);
+
+        if (unlikely(err)) {
+            pr_err("skcipher_walk_done() failed: %d\n", err);
+            return err;
+        }
+    }
+
+    return err;
+}
+
+static struct skcipher_alg xtsAesAlg = {
+    .base.cra_name          = WOLFKM_AESXTS_NAME,
+    .base.cra_driver_name   = WOLFKM_AESXTS_DRIVER,
+    .base.cra_priority      = WOLFSSL_LINUXKM_LKCAPI_PRIORITY,
+    .base.cra_blocksize     = AES_BLOCK_SIZE,
+    .base.cra_ctxsize       = sizeof(struct km_AesXtsCtx),
+    .base.cra_module        = THIS_MODULE,
+
+    .min_keysize            = 2 * AES_128_KEY_SIZE,
+    .max_keysize            = 2 * AES_256_KEY_SIZE,
+    .ivsize                 = AES_BLOCK_SIZE,
+    .walksize               = 2 * AES_BLOCK_SIZE,
+    .init                   = km_AesXtsInit,
+    .exit                   = km_AesXtsExit,
+    .setkey                 = km_AesXtsSetKey,
+    .encrypt                = km_AesXtsEncrypt,
+    .decrypt                = km_AesXtsDecrypt
+};
+static int xtsAesAlg_loaded = 0;
+
+#endif /* WOLFSSL_AES_XTS &&
+        * (LINUXKM_LKCAPI_REGISTER_ALL || LINUXKM_LKCAPI_REGISTER_AESXTS)
+        */
+
+/* cipher tests, cribbed from test.c, with supplementary LKCAPI tests: */
+
+#if defined(HAVE_AES_CBC) && \
+    (defined(LINUXKM_LKCAPI_REGISTER_ALL) || defined(LINUXKM_LKCAPI_REGISTER_AESCBC))
+
+static int linuxkm_test_aescbc(void)
+{
+    int     ret = 0;
+    struct crypto_skcipher *  tfm = NULL;
+    struct skcipher_request * req = NULL;
+    struct scatterlist        src, dst;
+    Aes     aes;
+    WOLFSSL_SMALL_STACK_STATIC const byte key32[] =
+    {
+        0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37,
+        0x38, 0x39, 0x61, 0x62, 0x63, 0x64, 0x65, 0x66,
+        0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37,
+        0x38, 0x39, 0x61, 0x62, 0x63, 0x64, 0x65, 0x66
+    };
+    WOLFSSL_SMALL_STACK_STATIC const byte vector[] = /* Now is the time for all good men w/o trailing 0 */
```

Comment:
Why not a KAT (Known Answer Test)?

Reply:
The FIPS KATs also get run, in a FIPS build, but @philljj adapted the `test.c` tests to assure reasonable self-test coverage even in non-FIPS builds.  I followed suit when I wrote the AES-XTS glue.

Reply:
The idea when I wrote this was a sanity test at startup to verify that calls through the kernel produce the same results as direct calls through our API.

Maybe a known answer test would be better.

Reply:
The encrypt and decrypt result are always the same.
Just make it a KAT rather than putting it through another API.

Reply:
OIC!  whoops, I was focused on AES-XTS, where I did indeed bring over full KATs from `test.c`.  the CBC/CFB/GCM tests indeed are not KATs.  will fix forthwith.

Reply:
fixed.